File: exceltest_ruby/exceltoword.rb
require "creek"
require "openxml/docx"

def word(item, file)
  package = OpenXml::Docx::Package.new
  include OpenXml::Docx::Elements
  line = item.split("\n")
  line.each do |lorem|
    text = Text.new(lorem)
    run = Run.new
    run << text
    paragraph = Paragraph.new
    paragraph << run
    package.document << paragraph
  end
  section_properties = SectionProperties.new
  font_size = OpenXml::Docx::Properties::FontSize.new(19)
  font_name = OpenXml::Docx::Properties::Font.new("宋体")
  section_properties << font_size
  section_properties << font_name
  package.document << section_properties
  package.save(file)
end

def joinlist(item, len)
  letters = ["B", "C", "D", "E", "F", "G", "H"]
  item[0] = item[0].to_s.delete("\n") + "\n"
  if item[1] && item[1].to_s.strip != ""
    item[1] = "A." + item[1].to_s.delete("\n")
  else
    item[1] = ""
  end
  if len == 3
    item[len - 1] = " 答案：" + item[len - 1].to_s.delete("\n") + "\n"
    return item.join
  end
  if len > 3
    for i in 0...(len - 3)
      if item[i + 2] && item[i + 2].to_s.strip != ""
        item[i + 2] = " " + letters[i] + "." + item[i + 2].to_s.delete("\n")
      else
        item[i + 2] = ""
      end
    end
    item[len - 1] = " 答案：" + item[len - 1].to_s.delete("\n") + "\n"
    return item.join
  end
end

def process(items)
  questionindex = []
  question = []
  answer = []
  choice = []
  for item in items
    if /题目/ =~ item
      question.push(items.index(item))
    end
    if /[ABCDEFGH]/i =~ item
      choice.push(items.index(item))
    end
    if /答案/ =~ item
      answer.push(items.index(item))
    end
  end
  if question.any?
    questionindex.push(question[0])
  end
  if choice.any?
    questionindex.concat(choice)
  end
  if answer.any?
    questionindex.push(answer[0])
  end
  if questionindex.any?
    return questionindex, questionindex.length
  end
  return nil, nil
end

def getcontentofsheet(sheet)
  temp = []
  sheet.rows.each_slice(1) do |row|
    temp.push(row[0].values)
  end
  if temp.any?
    return temp.length, temp[0], temp[0].length, temp
  end
  return nil, nil, nil, nil
end

data = Creek::Book.new "safety.xlsx"
sheetnumber = data.sheets.length
total = Array.new(sheetnumber)
table = Array.new(sheetnumber)
nrows = Array.new(sheetnumber)
ncols = Array.new(sheetnumber)
sum = Array.new(sheetnumber)

for i in 0...sheetnumber
  table[i] = data.sheets[i]
  nrows[i], firstrow, ncols[i], content = getcontentofsheet(table[i])
  if content
    sum[i] = []
    temp, fuck = process(firstrow)
    if temp
      for j in 0...(nrows[i] - 1)
        everyrow = []
        index = "(" + (j + 1).to_s + ")"
        for k in temp
          everyrow.push(content[j + 1][k.to_i])
        end
        everyrowlist = index + joinlist(everyrow, fuck)
        sum[i].push(everyrowlist)
      end
      total[i] = sum[i].join
      title = table[i].name.to_s + "Ruby.docx"
      word(total[i], title)
    end
  end
end

File: exceltest_python/exceltoword.py
from posixpath import split
import xlrd
from docx import Document
from docx.shared import Pt
from docx.oxml.ns import qn
import re

data = xlrd.open_workbook("safety.xlsx")
sheetsnumber = data.nsheets


def word(item, file):
    doc1 = Document()
    doc1.styles['Normal'].font.name = u'宋体'
    doc1.styles['Normal']._element.rPr.rFonts.set(qn('w:eastAsia'), u'宋体')
    doc1.styles['Normal'].font.size = Pt(9)
    doc1.add_paragraph(''.join(item))
    doc1.save(file)


def joinlist(item, fuck):
    letters = ['B', 'C', 'D', 'E', 'F', 'G', 'H']
    item[0] = str(item[0]).replace("\n", "") + "\n"
    item[1] = ("A." + str(item[1]).replace("\n", "") if (item[1])
               and ("".join(str(item[1]).split()) != "") else "")
    if fuck == 3:
        item[fuck - 1] = " 答案：" + str(item[fuck - 1]).replace("\n", "") + "\n"
        return ''.join(item)
    if fuck > 3:
        for i in range(fuck - 3):
            item[2+i] = (" " + str(letters[i]) + "." + str(item[2+i]).replace("\n", "")
                         if (item[2+i]) and ("".join(str(item[2+i]).split()) != "") else "")
        item[fuck - 1] = " 答案：" + str(item[fuck - 1]).replace("\n", "") + "\n"
        return ''.join(item)


def process(items):
    questionindex = []
    question = []
    answer = []
    choice = []
    modelregexone = re.compile(r'题目')
    modelregextwo = re.compile(r'答案')
    modelregexthree = re.compile(r'A|B|C|D|E|F|G|H', re.I)
    for item in items:
        if modelregexone.search(item):
            question.append(items.index(item))
        if modelregextwo.search(item):
            answer.append(items.index(item))
        if modelregexthree.search(item):
            choice.append(items.index(item))
    if len(question) != 0:
        questionindex.append(question[0])
    if len(choice) != 0:
        questionindex.extend(choice)
    if len(answer) != 0:
        questionindex.append(answer[0])
    if len(questionindex) != 0:
        return questionindex, len(questionindex)
    return False, False


total = [0 for i in range(sheetsnumber)]
table = [0 for i in range(sheetsnumber)]
nrows = [0 for i in range(sheetsnumber)]
ncols = [0 for i in range(sheetsnumber)]
sum = [0 for i in range(sheetsnumber)]
for i in range(sheetsnumber):
    table[i] = data.sheets()[i]
    nrows[i] = table[i].nrows
    ncols[i] = table[i].ncols
    if nrows[i] != 0:
        temp, fuck = process(table[i].row_values(
            0, start_colx=0, end_colx=ncols[i]))
        if temp:
            sum[i] = []
            for j in range(nrows[i]-1):
                everyrow = []
                index = "(" + str(j+1) + ")"
                for k in temp:
                    everyrow.append(table[i].cell_value(j+1, k))
                everyrowlist = index + joinlist(everyrow, fuck)
                sum[i].append(everyrowlist)
            total[i] = ''.join(sum[i])
            title = str(table[i].name) + "Python.docx"
            word(total[i], title)
